File: 프로그래머스/unrated/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
dx = [1,-1,0,0]
dy = [0,0,1,-1]

def solution(land):
    answer = 0
    n, m = len(land), len(land[0])
    columns = [0 for i in range(m)]
    visit = [[False for i in range(m)] for j in range(n)]
    
    for i in range(n):
        for j in range(m):
            if land[i][j] == 1 and not visit[i][j]:
                # find count
                visit[i][j] = True
                queue = [(i, j)]
                pos = [(i, j)]
                count = 1

                while queue:
                    y, x = queue.pop()

                    for idx in range(4):
                        ny, nx = dy[idx] + y, dx[idx] + x

                        if 0 <= ny < n and 0 <= nx < m:
                            if land[ny][nx] == 1 and not visit[ny][nx]:
                                visit[ny][nx] = True    
                                count += 1  
                                queue.append((ny, nx))
                                pos.append((ny, nx))
                
                cols = set([b for a, b in pos])
                
                # add
                for idx in cols:
                    columns[idx] += count

    return max(columns)

File: 프로그래머스/unrated/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;

class Solution {
    class Node{
        int y;
        int x;
        
        public Node(int y, int x){
            this.y = y;
            this.x = x;
        }
    }
    public int solution(int[][] land) {
        int answer = 0;
        int[] dy = new int[]{1, -1, 0, 0};
        int[] dx = new int[]{0, 0, 1, -1};
        int n = land.length;
        int m = land[0].length;
        boolean[][] visit = new boolean[n][m];
        int[] columns = new int[m];
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(land[i][j] == 0) continue;
                if(visit[i][j]) continue;
                
                visit[i][j] = true;
                int count = 1;
                Queue<Node> queue = new LinkedList<Node>();
                queue.add(new Node(i, j));
                Set<Integer> cols = new HashSet<>();
                cols.add(j);
                
                while(!queue.isEmpty()){
                    Node cur = queue.poll();
                    
                    for(int k = 0; k < 4; k++){
                        int ny = cur.y + dy[k];
                        int nx = cur.x + dx[k];
                        
                        if(ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
                        if(visit[ny][nx] || land[ny][nx] == 0) continue;
                        
                        visit[ny][nx] = true;
                        queue.add(new Node(ny, nx));
                        count++;
                        cols.add(nx);
                    }
                }
                
                for(Integer idx : cols) {
                    columns[idx] += count;
                }
            }
        }
        
        
        return Arrays.stream(columns).max().getAsInt();
    }
}